services/src/main/java/org/keycloak/authentication/forms/RegistrationProfile.java

```diff
@@ -57,41 +63,25 @@ public String getHelpText() {
     }
 
     @Override
-    public void validate(ValidationContext context) {
+    public void validate(org.keycloak.authentication.ValidationContext context) {
         MultivaluedMap<String, String> formData = context.getHttpRequest().getDecodedFormParameters();
-        List<FormMessage> errors = new ArrayList<>();
+        UserAttributesRepresentation rep = AttributeFormDataProcessor.toAttributesRepresentation(formData);
 
-        context.getEvent().detail(Details.REGISTER_METHOD, "form");
-        String eventError = Errors.INVALID_REGISTRATION;
-
-        if (Validation.isBlank(formData.getFirst((RegistrationPage.FIELD_FIRST_NAME)))) {
-            errors.add(new FormMessage(RegistrationPage.FIELD_FIRST_NAME, Messages.MISSING_FIRST_NAME));
-        }
-
-        if (Validation.isBlank(formData.getFirst((RegistrationPage.FIELD_LAST_NAME)))) {
-            errors.add(new FormMessage(RegistrationPage.FIELD_LAST_NAME, Messages.MISSING_LAST_NAME));
-        }
+        UserProfileProvider userProfile = context.getSession().getProvider(UserProfileProvider.class, LegacyUserProfileProvider.PROVIDER_ID);
 
-        String email = formData.getFirst(Validation.FIELD_EMAIL);
-        boolean emailValid = true;
-        if (Validation.isBlank(email)) {
-            errors.add(new FormMessage(RegistrationPage.FIELD_EMAIL, Messages.MISSING_EMAIL));
-            emailValid = false;
-        } else if (!Validation.isEmailValid(email)) {
-            context.getEvent().detail(Details.EMAIL, email);
-            errors.add(new FormMessage(RegistrationPage.FIELD_EMAIL, Messages.INVALID_EMAIL));
-            emailValid = false;
-        }
-
-        if (emailValid && !context.getRealm().isDuplicateEmailsAllowed() && context.getSession().users().getUserByEmail(email, context.getRealm()) != null) {
-            eventError = Errors.EMAIL_IN_USE;
-            formData.remove(Validation.FIELD_EMAIL);
-            context.getEvent().detail(Details.EMAIL, email);
-            errors.add(new FormMessage(RegistrationPage.FIELD_EMAIL, Messages.EMAIL_EXISTS));
-        }
+        context.getEvent().detail(Details.REGISTER_METHOD, "form");
+        UserProfileValidationResult result = userProfile.validate(rep, null, ValidationScenario.RegistrationProfile);
+        List<FormMessage> errors = Validation.getFormErrorsFromValidation(result);
 
         if (errors.size() > 0) {
-            context.error(eventError);
+            if (result.hasFailureOfErrorType(Messages.EMAIL_EXISTS, Messages.INVALID_EMAIL))
+                context.getEvent().detail(Details.EMAIL, rep.firstAttribute(UserModel.EMAIL));
+
+            if (result.hasFailureOfErrorType(Messages.EMAIL_EXISTS)) {
+                context.error(Errors.EMAIL_IN_USE);
+                formData.remove("email");
+            } else
```

Comment:
I think looking for specific error messages like this is too brittle, perhaps it would be better to have a `UserProfileValidationException` with sub-classes like `EmailInUseException` instead here?

Reply:
I would count that into the category of improvements we could do also later as the given behavior is quite near to what was there before. But I can also put it in this PR.